staging/vhost-device-sound/src/audio_backends/alsa.rs

```
@@ -114,17 +113,63 @@ fn update_pcm(
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_IMA_ADPCM => 2,
+                    virtio_sound::VIRTIO_SND_PCM_FMT_S18_3
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_U18_3
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_S20_3
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_U20_3
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_S24_3
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_U24_3
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_S24
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_U24 => 3,
+                    virtio_sound::VIRTIO_SND_PCM_FMT_S20
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_U20
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_S32
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_U32
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_FLOAT
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_DSD_U32
+                    | virtio_sound::VIRTIO_SND_PCM_FMT_IEC958_SUBFRAME => 4,
+                    virtio_sound::VIRTIO_SND_PCM_FMT_FLOAT64 => 8,
+                    _ => unreachable!(),
+                };
+
+            // Calculate desirable bytes/sec rate to achieve the stream's desired
+            // parameters:
+
+            let bps_rate = frame_size * rate;
+
+            // Calculate period size for ~100ms (arbitrary) interrupt period:
+
+            let period_bytes = bps_rate / 10;
+
+            // Finally, calculate the size of a period (in frames):
+
+            let period_frames = period_bytes / frame_size;
+
+            hwp.set_period_size(period_frames as i64, alsa::ValueOr::Less)?;
+
+            // Online ALSA driver recommendations seem to be that the buffer should be at
+            // least 2 * period_size
+            if let Err(err) = hwp.set_buffer_size_near(2 * period_frames as i64) {
+                log::error!("could not set buffer size {}: {}", 2 * period_frames, err);
```

Should we also return an error to the guest?
(I don't know alsa library, so maybe it is not the case)

---

I read on some docs that alsa sets buffer size itself, I think I should find an certain source for that claim and add it in the comments